torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -234,8 +236,64 @@ inline void errorIfCapturingNonCapturableNCCL() {
   }
 }
 
+void markRankNotStuck(char *rankStatusHelper, int rank) {
+  if (!rankStatusHelper) {
+    LOG(ERROR) << "rankStatusHelper is NULL";
+    return;
+  }
+  rankStatusHelper[rank] = 0x00;
+}
+
+void markRankStuck(char *rankStatusHelper, int rank) {
+  if (!rankStatusHelper) {
+    LOG(ERROR) << "rankStatusHelper is NULL";
+    return;
+  }
+  rankStatusHelper[rank] = 0x01;
+}
+
+void markRankTimedOut(char *rankStatusHelper, int rank) {
+  if (!rankStatusHelper) {
+    LOG(ERROR) << "rankStatusHelper is NULL";
+    return;
+  }
+  rankStatusHelper[rank] = 0x02;
+}
+
 } // namespace
 
+RankStatusHelper::RankStatusHelper(size_t worldSize): worldSize_(worldSize) {
+  // assume file exists
+  int fd;
+  if ((fd = open("rank_statuses", O_RDWR, 0)) == -1) {
```

Comment:
This is pretty fragile since there could be leftover files from previously failed runs. Can we instead create a temporary file and log that or create a file with the pid as a suffix to avoid collisions (pids can be reused so this might break too)?

Reply:
I truncate the file if it exists in `_create_rank_statuses_file` in `spawn.py` Does this suffice?